File: accsaber-api/src/main/java/de/ixsen/accsaber/api/error/AccsaberExceptionHandler.java
package de.ixsen.accsaber.api.error;

import com.auth0.jwt.exceptions.JWTVerificationException;
import com.auth0.jwt.exceptions.TokenExpiredException;
import de.ixsen.accsaber.business.HasLogger;
import de.ixsen.accsaber.business.exceptions.AccsaberOperationException;
import de.ixsen.accsaber.business.exceptions.ExceptionType;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

@ControllerAdvice
public class AccsaberExceptionHandler extends ResponseEntityExceptionHandler implements HasLogger {

    @ExceptionHandler(TokenExpiredException.class)
    public ResponseEntity<ErrorDto> handleInvalidToken(JWTVerificationException exception) {
        ErrorDto errorDto = new ErrorDto();
        errorDto.setErrorCode(ExceptionType.AUTH_INVALID.getErrorCode());
        errorDto.setMessage("Invalid authentication");

        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorDto);
    }

    @ExceptionHandler(AccsaberOperationException.class)
    public ResponseEntity<ErrorDto> handleAccsaberOperationException(AccsaberOperationException exception) {
        this.getLogger().warn(exception.getMessage());
        ErrorDto errorDto = new ErrorDto();
        errorDto.setErrorCode(exception.getErrorCode());
        errorDto.setMessage(exception.getMessage());

        return ResponseEntity.status(this.getStatus(exception.getErrorCode())).body(errorDto);
    }

    private HttpStatus getStatus(String errorCode) {
        switch (errorCode) {
            case "00100":
            case "00101":
                return HttpStatus.NOT_FOUND;
            default:
                return HttpStatus.BAD_REQUEST;
        }
    }

}

File: accsaber-api/src/main/kotlin/de/ixsen/accsaber/api/error/AccsaberExceptionHandler.kt
package de.ixsen.accsaber.api.error

import com.auth0.jwt.exceptions.JWTVerificationException
import com.auth0.jwt.exceptions.TokenExpiredException
import de.ixsen.accsaber.business.HasLogger
import de.ixsen.accsaber.business.exceptions.AccsaberOperationException
import de.ixsen.accsaber.business.exceptions.ExceptionType
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@ControllerAdvice
class AccsaberExceptionHandler : ResponseEntityExceptionHandler(), HasLogger {
    @ExceptionHandler(TokenExpiredException::class)
    fun handleInvalidToken(exception: JWTVerificationException?): ResponseEntity<ErrorDto> {
        val errorDto = ErrorDto()
        errorDto.errorCode = ExceptionType.AUTH_INVALID.errorCode
        errorDto.message = "Invalid authentication"
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorDto)
    }

    @ExceptionHandler(AccsaberOperationException::class)
    fun handleAccsaberOperationException(exception: AccsaberOperationException): ResponseEntity<ErrorDto> {
        logger.warn(exception.message)
        val errorDto = ErrorDto()
        errorDto.errorCode = exception.errorCode
        errorDto.message = exception.message
        return ResponseEntity.status(getStatus(exception.errorCode)).body(errorDto)
    }

    private fun getStatus(errorCode: String): HttpStatus {
        return when (errorCode) {
            "00100", "00101" -> HttpStatus.NOT_FOUND
            else -> HttpStatus.BAD_REQUEST
        }
    }
}